sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -207,9 +207,25 @@ case class InsertIntoHadoopFsRelationCommand(
     }
     // first clear the path determined by the static partition keys (e.g. /table/foo=1)
     val staticPrefixPath = qualifiedOutputPath.suffix(staticPartitionPrefix)
-    if (fs.exists(staticPrefixPath) && !committer.deleteWithJob(fs, staticPrefixPath, true)) {
-      throw new IOException(s"Unable to clear output " +
-        s"directory $staticPrefixPath prior to writing to it")
+
+    // check if delete the dir or just sub files
+    if (fs.exists(staticPrefixPath)) {
+      // check if is he table root, and record the file to delete
+      if (staticPartitionPrefix.isEmpty) {
+        val files = fs.listFiles(staticPrefixPath, false)
+        while (files.hasNext) {
+          val file = files.next()
+          if (!committer.deleteWithJob(fs, file.getPath, true)) {
```

Comment:
this deletes leaf files one by one, have you evaluated the performance difference?

Reply:
First of all, if it is the root directory of the table, I must record all the files in the directory, and wait until the job is commited to delete. Because the `_temporary` of the entire job is also in the directory, I cannot directly delete the entire directory.

Second, when we record the files that need to be deleted, we just list the files in the root directory non-recursively. Under normal circumstances, the number of files in the first-level directory of the partition table will not be too much.

In the end, this will certainly be slower than directly deleting the entire directory, but under the current implementation, we cannot directly delete the entire table directory.

Reply:
is it possible to only do it with overwrite? we should not introduce perf regression when not necessary.

Reply:
have you considered the approach taken by `dynamicPartitionOverwrite`? i.e. using staging directory.

Reply:
1. From the code point of view, the current implementation is `deleteMatchingPartitions` happend only if `overwrite` is specified.
2. Using `dynamicPartitionOverwrite` will not solve this problem,because it will also generate a `.stage` directory under the table root directory. We still need to record all the files we want to delete, but we cannot directly delete the root directories.
The dynamic partition overwrite is actually recording all the partitions that need to be deleted and then deleted one by one. And the entire table `overwrite` deletes all the data of the entire directory, it needs to record all deleted partition directory files,so in fact the implementation of the code is similar with `dynamicPartitionOverwrite` .

Reply:
> deleteMatchingPartitions happend only if overwrite is specified.

I mean, we should only do it if overwriting the same table.

Reply:
If I do this, when the job is committed, it will delete the entire `output` directory. And there will be no data.

Reply:
we will delete files just before committing job, do we?

Reply:
The key is that the data is already in the `output` directory before committing job, and we can't delete the `output` directory anymore.

We overloaded `FileCommitProtocol` in the `HadoopMapReduceCommitProtocol` with the `deleteWithJob` method. Now it will not delete the file immediately, but it will wait until the entire job is committed.

We did delete the files with committed the job, but the temporary output files were generated when the task was started. These temporary output files are in the `output` directory.  And the data will be move out to the `output` directory. 

After the job starts, there is no safe time to delete the entire `output` directory.

Reply:
That's a good idea. I change my code.